From bbe0479f7fd8cdcda4d0ec2b55f7421776f8b7c6 Mon Sep 17 00:00:00 2001
From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Date: Fri, 19 Apr 2024 00:42:57 +0900
Subject: [PATCH] ntdll: Properly track refcount on static imports of export
 forwarders.

Today this is only done for the the main EXE.  Generalize this to DLLs
as well.
---
 dlls/kernel32/tests/loader.c | 4 ++--
 dlls/ntdll/loader.c          | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/kernel32/tests/loader.c b/dlls/kernel32/tests/loader.c
index 1d7aa0dbb6f..3af04bbccb2 100644
--- a/dlls/kernel32/tests/loader.c
+++ b/dlls/kernel32/tests/loader.c
@@ -2839,7 +2839,7 @@ static void subtest_export_forwarder_dep_chain( size_t num_chained_export_module
 
         /* FreeLibrary() should *not* unload the DLL immediately */
         module = GetModuleHandleA( temp_paths[i] );
-        todo_wine_if(i < ultimate_depender_index && i + 1 != importer_index)
+        todo_wine_if(i < ultimate_depender_index && i + 1 != importer_index && !(i < importer_index))
         ok( module == modules[i], "modules[%Iu] expected %p, got %p (unloaded?) err=%lu\n",
             i, modules[i], module, GetLastError() );
     }
@@ -2851,7 +2851,7 @@ static void subtest_export_forwarder_dep_chain( size_t num_chained_export_module
     {
         HMODULE module = GetModuleHandleA( temp_paths[i] );
 
-        todo_wine_if(i < ultimate_depender_index && i + 1 != importer_index)
+        todo_wine_if(i < ultimate_depender_index && i + 1 != importer_index && !(i < importer_index))
         ok( module == modules[i], "modules[%Iu] expected %p, got %p (unloaded?) err=%lu\n",
             i, modules[i], module, GetLastError() );
     }
diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index e0c107052b1..6584a992fe0 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -954,7 +954,7 @@ static FARPROC find_forwarded_export( HMODULE module, const char *forward, LPCWS
 
     if (wm->ldr.DdagNode != node_ntdll && wm->ldr.DdagNode != node_kernel32)
     {
-        if (!imports_fixup_done && importer)
+        if (importer)
         {
             /* Prepare for the callee stealing the reference */
             if (!wm_loaded && wm->ldr.LoadCount != -1) wm->ldr.LoadCount++;
-- 
GitLab

